### PR TITLE
rename global.json to token.json, support localization in markdown

### PR DESCRIPTION
**Pull request by superyyrrzz**
@chenkennt @vwxyzh @hellosnow @qinezh @ansyral

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.Engine/MarkdownProviders/DfmServiceProvider.cs [hunk 1]
```diff
@@ -13,16 +14,16 @@ public class DfmServiceProvider : IMarkdownServiceProvider
...
-            return new DfmService(parameters.BasePath);
+            return new DfmService(parameters.BasePath, parameters.Tokens);
         }
 
         private sealed class DfmService : IMarkdownService
         {
             private readonly DfmEngineBuilder _builder;
 
-            public DfmService(string baseDir)
+            public DfmService(string baseDir, IDictionary<string, object> tokens)
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/MarkdownProviders/DfmServiceProvider.cs]: use ImmutableDictionary?

File: src/Microsoft.DocAsCode.MarkdownLite/IMarkdownRenderer.cs [hunk 2]
```diff
@@ -22,6 +22,10 @@ public interface IMarkdownRenderer
         /// Get the <see cref="Options"/>.
         /// </summary>
         Options Options { get; }
+        /// <summary>
+        /// Get the tokens for localization
+        /// </summary>
+        IDictionary<string, object> Tokens { get; }
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.MarkdownLite/IMarkdownRenderer.cs]: remove, I think it is a part of renderer.

**Comment by vwxyzh** [comment]
⌚

File: src/Microsoft.DocAsCode.Build.Engine/TemplateProcessor.cs [hunk 3]
```diff
@@ -40,7 +40,7 @@ public TemplateProcessor(ResourceCollection resourceProvider, int maxParallelism
 
             _resourceProvider = resourceProvider;
             _templateCollection = new TemplateCollection(resourceProvider, maxParallelism);
-            DefaultGlobalVariables = LoadGlobalJson(resourceProvider) ?? new Dictionary<string, object>();
+            DefaultGlobalVariables = LoadTokenJson(resourceProvider) ?? new Dictionary<string, object>();
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/TemplateProcessor.cs]: also rename?

File: src/docfx.website.themes/default/token.json [hunk 4]
```diff
@@ -34,5 +34,10 @@
   "eventType": "Event Type",
   "exceptions": "Exceptions",
   "condition": "Condition",
-  "extensionMethods": "Extension Methods"
+  "extensionMethods": "Extension Methods",
+  "note": "<h5>Note</h5>",
```
Comment (hellosnow) [inline src/docfx.website.themes/default/token.json]: why include h5 in the original token value?

File: src/Microsoft.DocAsCode.Dfm/DfmRenderer.cs [hunk 5]
```diff
@@ -87,11 +87,21 @@ public override StringBuffer Render(IMarkdownRenderer renderer, MarkdownBlockquo
...
                     var noteToken = (DfmNoteBlockToken)splitToken.Token;
+                    
                     content += "<div class=\"";
                     content += noteToken.NoteType.ToUpper();
-                    content += "\"><h5>";
-                    content += noteToken.NoteType.ToUpper();
-                    content += "</h5>";
+                    content += "\">";
+                    object heading;
+                    if (renderer.Tokens.TryGetValue(noteToken.NoteType.ToLower(), out heading))
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Dfm/DfmRenderer.cs]: better check null?

File: src/Microsoft.DocAsCode.Dfm/DfmRenderer.cs [hunk 6]
```diff
@@ -87,11 +87,21 @@ public override StringBuffer Render(IMarkdownRenderer renderer, MarkdownBlockquo
                 else if (splitToken.Token is DfmNoteBlockToken)
                 {
                     var noteToken = (DfmNoteBlockToken)splitToken.Token;
+
```
Comment (qinezh) [inline src/Microsoft.DocAsCode.Dfm/DfmRenderer.cs]: remove spaces

File: src/docfx.website.themes/default/token.json (hunk 4 above)
Comment (superyyrrzz) [inline src/docfx.website.themes/default/token.json]: MardownRenderer doesn't know the proper heading level in page. We allow token.json in template to custommize it.

**Comment by superyyrrzz** [comment]
@vwxyzh @hellosnow @qinezh updated, please help review.

File: src/Microsoft.DocAsCode.Dfm/DfmRenderer.cs [hunk 7]
```diff
@@ -12,10 +13,16 @@ namespace Microsoft.DocAsCode.Dfm
 
     public class DfmRenderer : HtmlRenderer
     {
+        private readonly ImmutableDictionary<string, string> _tokens;
         private static readonly DocfxFlavoredIncHelper _inlineInclusionHelper = new DocfxFlavoredIncHelper();
         private static readonly DocfxFlavoredIncHelper _blockInclusionHelper = new DocfxFlavoredIncHelper();
         private static readonly DfmCodeExtractor _dfmCodeExtractor = new DfmCodeExtractor();
 
+        public DfmRenderer(ImmutableDictionary<string, string> tokens)
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Dfm/DfmRenderer.cs]: can it be optional parameter?

File: src/Microsoft.DocAsCode.MarkdownLite/MarkdownEngineBuilder.cs [hunk 8]
```diff
@@ -3,6 +3,7 @@
 
 namespace Microsoft.DocAsCode.MarkdownLite
 {
+    using System.Collections.Generic;
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.MarkdownLite/MarkdownEngineBuilder.cs]: remove?

File: src/Microsoft.DocAsCode.Dfm/DfmEngineBuilder.cs [hunk 9]
```diff
@@ -4,6 +4,7 @@
 namespace Microsoft.DocAsCode.Dfm
 {
     using System;
+    using System.Collections.Generic;
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Dfm/DfmEngineBuilder.cs]: remove?

File: src/Microsoft.DocAsCode.Dfm/DfmRenderer.cs (hunk 7 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Dfm/DfmRenderer.cs]: for example, OpDfmRenderer in op plugg-in is child of this class, which not pass tokens as parameters

---
In reply to: [67845126](https://github.com/dotnet/docfx/pull/403#discussion_r67845126) [](ancestors = 67845126)

Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Dfm/DfmRenderer.cs]: nice catch! should be optional parameter

**Comment by hellosnow** [comment]
:shipit:

**Comment by qinezh** [comment]
:shipit:

**Pull request merged**